doc/source/programs/gdalwarp.rst

```diff
@@ -456,6 +457,32 @@ Starting with GDAL 3.1, it is possible to use as output format a driver that
 only supports the CreateCopy operation. This may internally imply creation of
 a temporary file.
 
+.. _gdalwarp_nodata:
+
+Nodata / source validity mask handling
+--------------------------------------
+
+Invalid values in source pixels, either identified trough a nodata value
+metadata set on the source band, a mask band, an alpha band or the use of
+:option:`-srcnodata` will not be used in interpolation.
+The details of how it is taken into account depends on the resampling kernel:
+
+- for nearest resampling, for each target pixel, the coordinate of its center
+  is projected back to source coordinates and the source pixel containing that
+  coordinate is identified. If this source pixel is invalid, the target pixel
+  is considered as nodata.
+
+- for bilinear, cubic, cubicspline and lanczos, for each target pixel, the
+  coordinate of its center is projected back to source coordinates and the
+  source pixel is identified. Given that those resampling kernels have a non-null
+  kernel radius, it might be possible that there are invalid values in other
+  contributing source pixels. The weights used to take into account those invalid
+  values will be set to zero to ignore them.
```

Comment:
Does this mean there's no rescaling of the weights? Resampling (with gdal_translate) the following grid to 2x:

```
ncols        5
nrows        1
xllcorner    0
yllcorner    0
cellsize     2
NODATA_value -10000
1  2  -10000 3  4
```

yields

```
ncols        10
nrows        2
xllcorner    0.000000000000
yllcorner    0.000000000000
cellsize     1.000000000000
NODATA_value -10000
 1 1 2 2 2 3 3 3 4 4
 1 1 2 2 2 3 3 3 4 4
```

while `gdalwarp` produces:

```
ncols        10
nrows        2
xllcorner    0.000000000000
yllcorner    0.000000000000
cellsize     1.000000000000
NODATA_value -10000
 1 1 2 2 -10000 -10000 3 3 4 4
 1 1 2 2 -10000 -10000 3 3 4 4
```

So:
 - it looks like the weights might be rescaled in the `gdal_translate` case
 - there is a difference between `gdalwarp` and `gdal_translate` that doesn't appear to be covered in here (but that might be fine)

Reply:
> it looks (I haven't computed it manually though) like the weights might be rescaled in the gdal_translate case

not sure what you mean by rescaled. They are set to null for nodata pixels both at the numerator and denominator, "obviously", to get something consistent.

I believe the difference in behaviour between gdalwarp and gdal_translate is explained by the doc (note that they are difference between gdalwarp and gdal_translate/gdaladdo). For gdalwarp, a main source pixel is selected to decide if the target pixel is going to be nodata. So here the source pixel at -10000 is going to give to be replicated in the output.  For gdal_translate, it is going to be ignored and only the neighbour pixels at 2 and 3 are taken into account.
The differences in behavior is obviously not super desirable, but I have no idea if there's a correct universal way of proceeding in such situation.

Reply:
> not sure what you mean by rescaled. They are set to null for nodata pixels both at the numerator and denominator, "obviously", to get something consistent.

Yeah, that's what I meant. Rescaled so the output isn't lower because one of the input pixels was ignored.

> I believe the difference in behaviour between gdalwarp and gdal_translate is explained by the doc (note that they are difference between gdalwarp and gdal_translate/gdaladdo).

Yeah, I probably misread it it in some way. In my reading:

 - for `gdalwarp`: the docs say that for one output pixel, there is _one_ input pixel ("the coordinate of its center is projected back to source coordinates and the source pixel is identified"), but then they talk about multiple "contributing source pixels". So the weights of the ignored pixels are set to 0, and the rest of the pixels are used.
 - for `gdal_translate`, there's multiple input pixels and in one specific case (`bicubic` with too many invalid pixels), the output pixel might be set to no data.

So unless I'm missing something, `gdal_translate` should output no data in more cases than `gdalwarp` for `bicubic`.

Reply:
> For gdalwarp, a main source pixel is selected to decide if the target pixel is going to be nodata.

another way of saying ""the coordinate of its center is projected back to source coordinates and the source pixel is identified"

Reply:
I'm still somewhat unconvinced (is the "main" pixel the one `nearest` would pick?), but I'm going to stop being annoying on this topic :-).

Reply:
> is the "main" pixel the one `nearest` would pick?

yes, but I realize that the sentence "If this source pixel is invalid, the target pixel is considered as nodata" is missing for that case. Adding it